src/main/java/net/minecraftforge/common/data/ForgeLootTableProvider.java

```diff
@@ -0,0 +1,162 @@
+                lootConditions.set(i, MatchTool.builder(ItemPredicate.Builder.create().tag(to)).build());
+                found = true;
+            } else if (lootCondition instanceof Inverted) {
+                ILootCondition invLootCondition = getField(Inverted.class, (Inverted) lootCondition, 0);
+
+                if (invLootCondition instanceof MatchTool && checkMatchTool((MatchTool) invLootCondition, from)) {
+                    lootConditions.set(i, Inverted.builder(MatchTool.builder(ItemPredicate.Builder.create().tag(to))).build());
+                    found = true;
+                } else if (invLootCondition instanceof Alternative && findAndReplaceInAlternative((Alternative) invLootCondition, from, to)) {
+                    found = true;
+                }
+            }
+        }
+
+        return found;
+    }
+
+    private boolean findAndReplaceInParentedLootEntry(ParentedLootEntry entry, Item from, Tag<Item> to) {
+        LootEntry[] lootEntries = getField(ParentedLootEntry.class, entry, 0);
+        boolean found = false;
+
+        for (LootEntry lootEntry : lootEntries) {
+            if (findAndReplaceInLootEntry(lootEntry, from, to)) {
+                found = true;
+            }
+        }
+
+        return found;
+    }
+
+    private boolean findAndReplaceInLootEntry(LootEntry entry, Item from, Tag<Item> to) {
+        ILootCondition[] lootConditions = getField(LootEntry.class, entry, 0);
+        boolean found = false;
+
+        for (int i = 0; i < lootConditions.length; i++) {
+            if (lootConditions[i] instanceof Alternative && findAndReplaceInAlternative((Alternative) lootConditions[i], from, to)) {
+                found = true;
+            } else if (lootConditions[i] instanceof MatchTool && checkMatchTool((MatchTool) lootConditions[i], from)) {
+                lootConditions[i] = MatchTool.builder(ItemPredicate.Builder.create().tag(to)).build();
+                found = true;
+            }
+        }
+
+        return found;
+    }
+
+    private boolean findAndReplaceInAlternative(Alternative alternative, Item from, Tag<Item> to) {
+        ILootCondition[] lootConditions = getField(Alternative.class, alternative, 0);
+        boolean found = false;
+
+        for (int i = 0; i < lootConditions.length; i++) {
+            if (lootConditions[i] instanceof MatchTool && checkMatchTool((MatchTool) lootConditions[i], from)) {
+                lootConditions[i] = MatchTool.builder(ItemPredicate.Builder.create().tag(to)).build();
+                found = true;
+            }
+        }
+
+        return found;
+    }
+
+    private boolean checkMatchTool(MatchTool lootCondition, Item expected) {
+        ItemPredicate predicate = getField(MatchTool.class, lootCondition, 0);
+        Item item = getField(ItemPredicate.class, predicate, 4);
+        return item != null && item.equals(expected);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T, R> R getField(Class<T> clz, T inst, int index)
```

Comment:
Depending on field order doesn't feel very safe, is there a reason this is used over ObfuscationReflectionHelper or a simple Access Transformer?

Reply:
That's also what's happening in `ForgeRecipeProvider`, where obtaining the list or map of ingredients relies on the field ordering (why?). There should be a common consensus on the approach to follow.
IMO `ObfuscationReflectionHelper` should be preferred over ATs (since these data Gens don't run at runtime and we don't really need to expose data) and field ordering.

Reply:
I was using `ForgeRecipeProvider` as example, but I can replace that by method from `ObfuscationReflectionHelper`